### ISSUE TITLE
Incorrect string literal type deduction  when class implements an interface.

### ISSUE THREAD

**Issue by NStal** [Microsoft/TypeScript#14790]
**TypeScript Version:**  2.1.4

**Code**

```ts
interface RouteSpec{
    path:"/user"
}
class Route implements RouteSpec{
    path = "/user"
}
```

**Expected behavior:**
No error reported

**Actual behavior:**

```
error TS2420: Class 'Route' incorrectly implements interface 'RouteSpec'.
  Types of property 'path' are incompatible.
    Type 'string' is not assignable to type '"/user"'.

```

**Comment by NStal** [comment]
I suggest we deduce "/user" to type `string & "/user"`

**Comment by kitsonk** [comment]
The challenge here is partially of intent.  At a class level, you have assigned a value to a "variable" which is mutable.  You are not contracting that that variable is immutable in the definition, therefore the compiler is assigning it a type of `string`.  If you were to say that the value is immutable, the compiler would stop complaining.  For example, this does not error:

```typescript
interface RouteSpec{
    path:"/user"
}
class Route implements RouteSpec{
    readonly path = "/user"
}
```

**Comment by NStal** [comment]
That's cool, thank you so much.

**Comment by NStal** [comment]
```typescript
interface RouteSpec {
    deps: ["Another"]
}
class Route implements RouteSpec {
    readonly deps = ["Another"]
}
```
But still fail on this case. 

```
test.ts(6,7): error TS2420: Class 'Route' incorrectly implements interface 'RouteSpec'.
  Types of property 'deps' are incompatible.
    Type 'string[]' is not assignable to type '["Another"]'.
      Property '0' is missing in type 'string[]'.
```

**Comment by kitsonk** [comment]
First, tuples can never be contextually inferred (there are several issues about that and why that is).  Second, the contents of values that are passed by reference cannot be immutable, therefore there is no way to get them contextually inferred the values are literals.  There is not enough contextual information for the compiler to make the right decision.

In a class `readonly` is like declaring a `const` outside of class and the type inferral follows those rules.  So for example if you did this:

```typescript
const deps = ["Another"]; // will always be string[]
```

You could mean `["Another"]` or `[string]` or `string[]` or `("Another")[]`.

In these cases where it is ambiguous, you have to be specific.

```
interface RouteSpec {
    deps: ["Another"]
}
class Route implements RouteSpec {
    deps: ["Another"] = ["Another"]
}
```

**Comment by NStal** [comment]
Yeah, I can see the problem. Thanks for your explaination.
